### ISSUE TITLE
No/bad support for installing to a different prefix

### ISSUE THREAD

**Issue by VorpalBlade** [FedeDP/Clightd#13]
Installing to a non-standard directory (i.e. /opt/clight or /usr/local/bin) is poorly supported, since the systemd service file (and possibly other things, I haven't checked further) use hard coded paths.

This would be a very useful feature to support installing the program outside of package manager control (e.g. when the program is not yet packaged for the distribution you use, Debian in my case). It would also be required for any distribution using a different install scheme (such as NixOS).

The makefile also seems to be ignoring standard variables such as LDFLAGS.

I would fix this and do a pull request if I knew what the preferred solution was, personally I would simply switch to a standard build system that handles nuances like this, such as cmake or autotools, but given that you have avoided such solutions that I assume you would not want that.

**Comment by FedeDP** [comment]
Hi @VorpalBlade , thanks for the feedback.

I am not against cmake (in fact, i already use it on some other projects). Moreover, i must say that i already thought about that switch.
But as i haven't got much time to work on my side projects i always considered the porting to cmake a "nice idea, but later" thing, ie: very low priority.

If you think you're willing to take the job, i would really be grateful to you; it would really help and i'd appreciate that! :)

Btw tomorrow i intend to merge devel branch; after that it will be just bugfixing and new bus api documenting until 3.0 release, so it would be the perfect time to port it to cmake.

Thank you very much!

**Comment by FedeDP** [comment]
Here i am again; i have merged devel branch right now (together with a push on Clight to support new clightd api).
You can now work on this, if you are willing to :)

**Comment by VorpalBlade** [comment]
For my goals, it would also be necessary to patch up the systemd file with the proper path, presumably using configure_file() if using cmake. Maybe some other files too.

I will look into doing this over the next couple of weeks, my schedule is quite sporadic and unpredictable, so I'm not sure exactly when though.

**Comment by FedeDP** [comment]
I think i am not in a hurry, so take all the time you need!

**Comment by VorpalBlade** [comment]
Looking a bit further into the build system, I'm slightly confused as to the relation between clight and clightd. It seems I had incorrectly understood it as clight being a frontend talking to clightd, but that doesn't seem to be the case since there is no dependency there. And clightd seems to have an optional dependency on clight in AUR.

What exactly is the relationship here? It would be useful to know which to port to cmake first (or if that doesn't matter).

**Comment by FedeDP** [comment]
Clight needs Clightd (in fact clight's PKGBUILD has a dep on clightd); but it is a runtime dependency, not a build one. So i think the build system should not be aware of that.

Clightd does not need Clight: it is just a system service that provides its functionalities; anybody can write a "frontend" to it. Thus Clight is just a "suggested" dependency for Clightd.

**Comment by VorpalBlade** [comment]
While messing about with this I found a different issue, in that you include for example "module/modules_easy.h", but pkg-config returns:
```
$ pkg-config --cflags libmodule
-I/opt/clight/include/module
```
So for this to work properly, you would need to just include "modules_easy.h", or make pkg-config return /opt/clight/include in this case. The only reason that the make file builds for me is:
```
$ pkg-config --cflags ddcutil
-I/opt/clight/include -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include
```
So this indirectly gives me the path needed to build with libmodule as well.

**Comment by FedeDP** [comment]
You are right...to be honest i was aware that i could/should use only `#include "module_eash.h"`; but in my standard path (/usr/include/) it was the same, so i left it there, without thinking for non-standard installations.

So, my fault! Are you willing to make a separate patch to fix libmodule includes?

**Comment by FedeDP** [comment]
Btw you probably noted that travis support is quite broken for clight and clightd. It is caused by https://github.com/mikkeloscar/arch-travis/issues/49.  
I hoped it would get fixed quickly; instead my builds are broken since a month or so.

**Comment by FedeDP** [comment]
> So for this to work properly, you would need to just include "modules_easy.h"

Fixed in https://github.com/FedeDP/Clightd/commit/f8fe3376043e0cf722cd3b63f2411198ee6a42f8.

**Comment by VorpalBlade** [comment]
Hm, I was thinking that the proper way might be to change libmodule such that you have to include with module/, but perhaps not. As for making a patch, you beat me to it (blame time zones). And I did indeed notice that travis seemed broken, and that an email (your?) is hard coded into the travis.yml file (so presumably you will get emails about failed builds in my fork of the repository?).

**Comment by FedeDP** [comment]
> Hm, I was thinking that the proper way might be to change libmodule such that you have to include with module/

I was wondering the same this morning... I think i'll take a better look at it.

> As for making a patch, you beat me to it (blame time zones).

:D

> And I did indeed notice that travis seemed broken, and that an email (your?) is hard coded into the travis.yml file (so presumably you will get emails about failed builds in my fork of the repository?).  

Yes, it is hardcoded there. Reading here: https://docs.travis-ci.com/user/notifications/#configuring-email-notifications it seems it can be dropped entirely and a notifcation will be sent to commit author.
I'll have a look at this.
Thanks!

**Comment by FedeDP** [comment]
> And I did indeed notice that travis seemed broken, and that an email (your?) is hard coded into the travis.yml file (so presumably you will get emails about failed builds in my fork of the repository?).

Fixed.

> Hm, I was thinking that the proper way might be to change libmodule such that you have to include with module/

Fixed in latest libmodule (still needs to adapt Clightd though).

**Comment by VorpalBlade** [comment]
Got some more time to work on this today, and I'm changing -DUSE_DDC to -DDDC_PRESENT (to be consistent with the other optional dependencies), makes it easier to define a helper macro for optional dependencies (because pkg-config doesn't play very nice with cmake when libraries are installed to non-standard paths).

**Comment by VorpalBlade** [comment]
Something I'm slightly confused by is the whole /usr/share/licenses thing. Is that an Arch Linux thing? I don't have it on either Debian nor Ubuntu (the only two distros that I use these days, though I used to use Arch, Gentoo and a few others a long time ago). I do have a lot of license files in /usr/share/doc/<package-name> though.

Might be worth looking into what the standard way here is supposed to be.

**Comment by FedeDP** [comment]
> Something I'm slightly confused by is the whole /usr/share/licenses thing. Is that an Arch Linux thing

I don't really know if this is an archlinux thing; i mean, /usr/share/licenses/ makes much more sense of /usr/share/doc/ for licenses to me.  
I just checked a couple of arch official packages' PKGBUILDs and i see that nobody uses a different "make install" for license, ie: it seems all these packages install their license in /usr/share/licenses.
I think we can avoid installing license file for now and decide later.
I could not find any xdg specification for it though.

**Comment by FedeDP** [comment]
Closed with #1d4187c31befc88ef39a11b285bd1dc7a1a37fde.
